### PR TITLE
Add TODO comments check to pylint

### LINKED ISSUE #992: Verify if in-code TODOs have a corresponding issue
### Feature Request

Would be great to have a tool (an action?) that runs from time to time and checks the code for TODO's with issue numbers

### PR DESCRIPTION
**Pull request by tomek0123456789**
Closes #992 

## Introduced changes

- Added custom rule to pylint to check for invalid TODO comments
Error looks like this:
```
Poe => pylint starknet_py
************* Module starknet_py.net.schemas.common
starknet_py/net/schemas/common.py:30:0: E2137: Uses a "TODO" comment without specifying issue number. (todo-issue-error)
************* Module starknet_py.proxy.contract_abi_resolver
starknet_py/proxy/contract_abi_resolver.py:97:0: E2137: Uses a "TODO" comment without specifying issue number. (todo-issue-error)
starknet_py/proxy/contract_abi_resolver.py:126:0: E2137: Uses a "TODO" comment without specifying issue number. (todo-issue-error)
************* Module starknet_py.tests.e2e.declare.declare_test
starknet_py/tests/e2e/declare/declare_test.py:22:0: E2137: Uses a "TODO" comment without specifying issue number. (todo-issue-error)
************* Module starknet_py.tests.e2e.docs.guide.test_cairo1_contract
starknet_py/tests/e2e/docs/guide/test_cairo1_contract.py:14:0: E2137: Uses a "TODO" comment without specifying issue number. (todo-issue-error)
************* Module starknet_py.tests.e2e.client.client_test
starknet_py/tests/e2e/client/client_test.py:523:0: E2137: Uses a "TODO" comment without specifying issue number. (todo-issue-error)
starknet_py/tests/e2e/client/client_test.py:542:0: E2137: Uses a "TODO" comment without specifying issue number. (todo-issue-error)
starknet_py/tests/e2e/client/client_test.py:568:0: E2137: Uses a "TODO" comment without specifying issue number. (todo-issue-error)
starknet_py/tests/e2e/client/client_test.py:597:0: E2137: Uses a "TODO" comment without specifying issue number. (todo-issue-error)
starknet_py/tests/e2e/client/client_test.py:627:0: E2137: Uses a "TODO" comment without specifying issue number. (todo-issue-error)

------------------------------------------------------------------
Your code has been rated at 9.91/10 (previous run: 9.91/10, +0.00)
```
`starknet_py/tests/e2e/client/client_test.py:523:0` comment:
```
    # TODO: Replace with `client` when RPC 0.3.0 is supported
```

##

- [ ] This PR contains breaking changes

### CODE REVIEW

File: pylint_todo_checker.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    """This required method auto registers the checker during initialization.
+    :param linter: The linter to register the checker to.
+    """
+    linter.register_checker(TodoTokenChecker(linter))
+
+
+class TodoTokenChecker(checkers.BaseChecker):
+    __implements__ = interfaces.ITokenChecker
+
+    name = "todo-error"
```
Comment (war-in) [inline pylint_todo_checker.py:17]:
```suggestion
    name = "todo-issue-error"
```

File: pylint_todo_checker.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    priority = -1
+    msgs = {
+        "E2137": (
+            'Uses a "TODO" comment without specifying issue number.',
+            "todo-error",
```
Comment (war-in) [inline pylint_todo_checker.py:22]:
```suggestion
            self.name,
```
?

File: pylint_todo_checker.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    :param linter: The linter to register the checker to.
+    """
+    linter.register_checker(TodoTokenChecker(linter))
+
+
+class TodoTokenChecker(checkers.BaseChecker):
+    __implements__ = interfaces.ITokenChecker
+
+    name = "todo-error"
+    priority = -1
```
[the remark this answers is not in the document]
Comment (tomek0123456789) [reply pylint_todo_checker.py:18]: > priority: A negative integer indicating when the check should run. Checkers run from the most negative priority to most positive.

as stated [here](https://github.com/oppia/oppia/wiki/Custom-Pylint-checks).

File: pylint_todo_checker.py (hunk 2 above)
Comment (tomek0123456789) [reply pylint_todo_checker.py:22]: `name` works, thanks!

**Comment by tomek0123456789** [comment]
Error looks like this:
```
Poe => pylint starknet_py
************* Module starknet_py.net.schemas.common
starknet_py/net/schemas/common.py:30:0: E2137: Uses a "TODO" comment without specifying issue number. (todo-issue-error)
************* Module starknet_py.proxy.contract_abi_resolver
starknet_py/proxy/contract_abi_resolver.py:97:0: E2137: Uses a "TODO" comment without specifying issue number. (todo-issue-error)
starknet_py/proxy/contract_abi_resolver.py:126:0: E2137: Uses a "TODO" comment without specifying issue number. (todo-issue-error)
************* Module starknet_py.tests.e2e.declare.declare_test
starknet_py/tests/e2e/declare/declare_test.py:22:0: E2137: Uses a "TODO" comment without specifying issue number. (todo-issue-error)
************* Module starknet_py.tests.e2e.docs.guide.test_cairo1_contract
starknet_py/tests/e2e/docs/guide/test_cairo1_contract.py:14:0: E2137: Uses a "TODO" comment without specifying issue number. (todo-issue-error)
************* Module starknet_py.tests.e2e.client.client_test
starknet_py/tests/e2e/client/client_test.py:523:0: E2137: Uses a "TODO" comment without specifying issue number. (todo-issue-error)
starknet_py/tests/e2e/client/client_test.py:542:0: E2137: Uses a "TODO" comment without specifying issue number. (todo-issue-error)
starknet_py/tests/e2e/client/client_test.py:568:0: E2137: Uses a "TODO" comment without specifying issue number. (todo-issue-error)
starknet_py/tests/e2e/client/client_test.py:597:0: E2137: Uses a "TODO" comment without specifying issue number. (todo-issue-error)
starknet_py/tests/e2e/client/client_test.py:627:0: E2137: Uses a "TODO" comment without specifying issue number. (todo-issue-error)

------------------------------------------------------------------
Your code has been rated at 9.91/10 (previous run: 9.91/10, +0.00)

```

**Comment by tomek0123456789** [comment]
Also, would like to get your opinion on what should be the format of TODO comments. Right now those ones are accepted: 
```
# TODO: (#number) todo comment
# TODO (#number) todo comment
```
but information about the fact that colon is optional is redundant IMO and we should stick to one.

File: pylint_todo_checker.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+class TodoTokenChecker(checkers.BaseChecker):
+    __implements__ = interfaces.ITokenChecker
+
+    name = "todo-issue-error"
+    priority = -1
+    msgs = {
+        "E2137": (
+            "Wrong TODO format, should be: '# TODO: (#issueno)...', colon optional.",
+            name,
+            "TODO comments should be in a format: '# TODO: (#issueno)...' (colon is optional).",
```
Comment (drknzz) [inline pylint_todo_checker.py:25]:
```suggestion
            "Invalid TODO format, should be: '# TODO: (#issue number) ...'.",
            name,
            "TODO comments should be in a format: '# TODO: (#issue number) ...'.",
```

**Comment by war-in** [comment]
> There is one problem though, pylint checks only in python files. There are two TODO comments in cairo1 contracts and they won't be caught by current solution.

It's perfectly fine for me. We do not want pylint to check cairo files

**Comment by war-in** [comment]
> but information about the fact that colon is optional is redundant IMO and we should stick to one option.

How about
`TODO (#number): ...?

**Pull request merged**